File: ros2/ros2_qos_examples_py/ros2_qos_examples_py/camera_publisher.py
"""Publishes a synthetic camera Image at 60 Hz with qos_profile_sensor_data."""

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image


class CameraPublisher(Node):
    def __init__(self) -> None:
        super().__init__('camera_publisher_py')
        self.pub = self.create_publisher(
            Image, 'camera/image', qos_profile_sensor_data)
        self.timer = self.create_timer(1.0 / 60.0, self._publish_frame)
        self.frame_id = 0

    def _publish_frame(self) -> None:
        msg = Image()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'camera'
        msg.height = 8
        msg.width = 8
        msg.encoding = 'mono8'
        msg.step = 8
        msg.data = bytes([self.frame_id & 0xFF] * 64)
        self.pub.publish(msg)

        if self.frame_id % 60 == 0:
            self.get_logger().info(f'published frame {self.frame_id}')
        self.frame_id += 1


def main() -> None:
    rclpy.init()
    rclpy.spin(CameraPublisher())
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2/ros2_qos_examples_cpp/src/camera_publisher.cpp
// Publishes a synthetic "camera" Image at 60 Hz using SensorDataQoS:
//   KEEP_LAST(5), BEST_EFFORT, VOLATILE.
// Pair with latest_frame_subscriber to observe drop behavior under a slow consumer.

#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"

class CameraPublisher : public rclcpp::Node
{
public:
  CameraPublisher() : Node("camera_publisher")
  {
    pub_ = create_publisher<sensor_msgs::msg::Image>(
      "camera/image", rclcpp::SensorDataQoS());

    timer_ = create_wall_timer(
      std::chrono::milliseconds(1000 / 60),
      [this]() { publish_frame(); });
  }

private:
  void publish_frame()
  {
    sensor_msgs::msg::Image msg;
    msg.header.stamp = now();
    msg.header.frame_id = "camera";
    msg.height = 8;
    msg.width = 8;
    msg.encoding = "mono8";
    msg.step = 8;
    msg.data.assign(64, static_cast<uint8_t>(frame_id_ & 0xFF));
    pub_->publish(msg);

    if ((frame_id_ % 60) == 0) {
      RCLCPP_INFO(get_logger(), "published frame %lu", frame_id_);
    }
    ++frame_id_;
  }

  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  uint64_t frame_id_{0};
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CameraPublisher>());
  rclcpp::shutdown();
  return 0;
}
